rosette/rbl/rosette/namespace-util.rbl

```diff
@@ -0,0 +1,77 @@
+;;; TODO: Remove unused methods in Tuplespace and add tests for the rest
+
+;;; TODO: Add tag for whether persistent
+;;; TODO: Translate match in compiler
+
+(random-number-init)
+;;; The debugSymbol is inserted so that the output is easier to read
+(defOprn generateFresh)
+(defProc (generateFresh debugSymbol)
+  (->symbol
+   (string-concat "Rholang-" debugSymbol "-" (->string (random-number)))))
+
+(defOprn append)
+(defPure Tuple (append item)
+  (concat (self) [item]))
+
+;;; Deletes element at position i in Tuple
+(defOprn delete-i)
+(defPure Tuple (delete-i p)
+  (concat (prim-sub-object (self) 0 p)
+          (prim-sub-object (self) (fx+ p 1)
+                           (fx- (prim-size (self)) (fx+ p 1)))))
+
+(defOprn symbol-list-concat)
+(defPure Tuple (symbol-list-concat) (->symbol (fold (self)
+                                                    (proc [e acc k]
+                                                      (k (string-concat acc (->string e)))) "")))
+
+;;; TODO: Figure out a more general flatten method
+(defOprn append-tuples)
+(defPure Tuple (append-tuples) (fold (self) (proc [e acc k] (k (tuple-concat acc e))) []))
```

Comment:
Is this different to
```
(defPure Tuple (append-tuples) (concat & (self)))
```
the built-in one-level flatten?

Reply:
Hmm I guess not?

```
*** no argument for dispatch: ((FOp  concat))
 in (FOp  append-tuples), on (FO TupleSBO 150769784)
```

Reply:
I see what is going on.

So the above implementation can't handle

(append-tuples [])

Reply:
Ok so

`(defPure Tuple (append-tuples) (if (null? (self)) [] (concat & (self))))`

works.

Reply:
```
Rosette System, Version 4.0.0 - Copyright 1989, 1990, 1991, 1992 MCC
 - Copyright 2017, 2018 RChain Cooperative
rosette> (defOprn append-tuples)
'append-tuples
rosette> (defPure Tuple (append-tuples) (concat & (self)))
(FOp  append-tuples)
rosette> (append-tuples [['a 'b] ['c 'd] ['e ['f 'g]]])
['a 'b 'c 'd 'e ['f 'g]]
```
Mine was doing the same as yours, so I pulled, rebuilt, and it worked.

(It actually didn't matter what I put for the body of the function; the oprn wasn't found.)

Reply:
That's a bit irritating. Why would `[]` receive the message? `[]` ain't a tuple. Anyhoo, maybe this is clearer:
```
rosette> (defPure Nil (append-tuples) [])
(FOp  append-tuples)
rosette> (append-tuples [])
[]
rosette> (append-tuples [['a 'b] ['c 'd] ['e ['f 'g]]])
['a 'b 'c 'd 'e ['f 'g]]
```

Reply:
Hmm @kirkwood I'm not quite clear what you last comment means. I'm planning to switch to `(defPure Tuple (append-tuples) (if (null? (self)) [] (concat & (self))))` (a modified version of what you proposed as the top comment) but that is ok with you right?

Reply:
Oh nvm. I didn't see your last message. I guess Github isn't for real time chat.

Reply:
Yeah, it's fine. My comment just pointed out that you can define the `[]` case by specializing on `Nil`, and leaving the rest the same. Not trying to bikeshed (this was all kicked off by a comment that was sitting unsent in my browser for days :))

Reply:
Yeah I definitely like special casing better. As for the reason that Tuple receives the message:

```
rosette> (parent (parent []))
(FO TupleSBO 145903736)
rosette> (parent [1 2 3])
(FO TupleSBO 145903736)
```